packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/bin/server/controllers/authentication_controller.dart

```diff
@@ -0,0 +1,94 @@
+import 'package:shelf/shelf.dart';
+
+import '../models/auth_token.dart';
+import '../repositories/auth_token_repository.dart';
+import '../utils/api_controller.dart';
+import '../utils/server_exceptions.dart';
+
+// ignore_for_file: cascade_invocations
+
+class AuthenticationController extends ApiController {
+  static const _authHeader = 'Authorization';
+  final _authTokenRepository = AuthTokenRepository();
+
+  @override
+  void registerRequests(WrappedRouter router) {
+    router.addRequest(
+      RequestType.POST,
+      '/api/authenticate',
+      _authenticationHandler,
+    );
+    router.addRequest(
+      RequestType.POST,
+      '/api/logout',
+      _logoutHandler,
+    );
+  }
+
+  bool isAuthenticated(Request request) {
+    final headers = request.headers;
+    if (!headers.containsKey(_authHeader)) {
+      throw UnauthorizedException('User not authorized!');
+    }
+
+    final accessToken = _getAccessTokenFromAuthHeader(request.headers);
+    if (!_validateAccessToken(accessToken)) {
+      throw UnauthorizedException('User not authorized!');
+    }
+
+    return true;
+  }
+
+  Future<Response> _authenticationHandler(Request request) async {
+    final params = await request.bodyFromFormData();
+    final refreshToken = params['refreshToken'];
+
+    if (refreshToken == null || refreshToken.isEmpty) {
+      if (params['username']?.isEmpty ?? true) {
+        throw BadRequestException('The username cannot be empty.');
+      }
+      if (params['password']?.isEmpty ?? true) {
+        throw BadRequestException('The password cannot be empty.');
+      }
+    }
+    final _token = _issueNewToken(refreshToken);
+    return responseBuilder.buildOK(data: _token.toJson());
+  }
+
+  Response _logoutHandler(Request request) {
+    isAuthenticated(request);
+
+    final accessToken = _getAccessTokenFromAuthHeader(request.headers);
+    _authTokenRepository.removeToken(accessToken);
+
+    return responseBuilder.buildOK();
+  }
+
+  String _getAccessTokenFromAuthHeader(Map<String, String> headers) {
+    try {
+      // Usually the auth header looks like 'Bearer token', but if the format
+      // is not respected, it may throw errors
+      return headers[_authHeader]?.split(' ')[1] ?? '';
+    } catch (e) {
+      return '';
+    }
+  }
+
+  bool _validateAccessToken(String accessToken) {
+    final token = _authTokenRepository.getToken(accessToken);
+    if (token == null) {
+      throw UnauthorizedException('Invalid access token!');
+    }
+    return token.isValid;
+  }
+
+  AuthToken _issueNewToken(String? refreshToken) {
+    if (refreshToken != null) {
+      final _tkn = _authTokenRepository.getTokenViaRefreshToken(refreshToken);
+      if (_tkn == null) throw BadRequestException('Invalid refresh token!');
+      _authTokenRepository.removeToken(_tkn.token);
+    }
+
+    return _authTokenRepository.issueNewToken();
+  }
```

Comment:
This method is just a convenience method that generates a new token (auth and refresh tokens). If the refresh token is provided, then it is removed from the database.

The refresh token should be used as a quick way of issuing a new token instead of letting the user re-log in. So only one of two possible auth scenarios can be applied at once:
1. The user authenticates with a refresh token (username and passwords are not checked at all). The refresh token bypasses the login credentials.
2. The user authenticates with username and password (since no refresh token exists).

Reply:
And what will happen if we call this method with refreshToken = null (it is nullable)

Reply:
Well, looking at the code it will issue a new token (containing auth and refresh tokens). Since no refresh token is passed (it cant remove it), it will just create a new one.

This scenario without a refresh token is triggered when the user authenticates with a username and password. Later, when authentication is required, the user refresh token will be passed along.

Reply:
I think that implementing the server side code we have to take care of server safety. I suppose we have to save the expected refreshToken and check if the received from the app is the same. And only if it is, to return a new token.

Reply:
Of course. I suggest taking another look at the method to clear things up.

If we take a look at [this line](https://github.com/Prime-Holding/rx_bloc/blob/feature/rx-bloc-cli-push-notifications/packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/bin/server/controllers/authentication_controller.dart#L87), we can see that we fetch the whole stored token from the token repository (based on the provided refresh token). If the refresh token does **NOT** exist, we return an error response in the next line. If it exists, we just remove the entry and issue a new one.

Reply:
Thank you that you turned my attention in this direction!